pkg/util/limiter/query_limiter.go

```diff
@@ -0,0 +1,89 @@
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"go.uber.org/atomic"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type QueryLimiter struct {
+	uniqueSeriesMx sync.RWMutex
+	uniqueSeries   map[model.Fingerprint]struct{}
+
+	chunkBytesCount   *atomic.Int32
+	maxSeriesPerQuery int
+}
+type queryLimiterCtxKey struct{}
+
+var (
+	qlCtxKey        = &queryLimiterCtxKey{}
+	errMaxSeriesHit = "The query hit the max number of series limit while fetching chunks %s (limit: %d)"
+)
+
+// NewQueryLimiter makes a new per-query rate limiter. Each per-query limiter
+// is configured using the `maxSeriesPerQuery` and `maxChunkBytesPerQuery` limits.
+func NewQueryLimiter(maxSeriesPerQuery int) *QueryLimiter {
+	return &QueryLimiter{
+		uniqueSeriesMx: sync.RWMutex{},
+		uniqueSeries:   map[model.Fingerprint]struct{}{},
+
+		chunkBytesCount: atomic.NewInt32(0),
+
+		maxSeriesPerQuery: maxSeriesPerQuery,
+	}
+}
+
+func NewQueryLimiterOnContext(ctx context.Context, maxSeriesPerQuery int) context.Context {
```

Comment:
Comment from previous review: do we need this function? We can do the same combining `AddQueryLimiterToContext()` and `NewQueryLimiter()` together.